### PR TITLE
[flang] AliasAnalysis: More formally define and distinguish between data and non-data 

### PR DESCRIPTION
**Pull request by Renaud-K**
This PR is an implementation for changes proposed in https://discourse.llvm.org/t/rfc-distinguish-between-data-and-non-data-in-fir-alias-analysis/78759

Test updates were made when the query was on the wrong reference. So, it is my hope that this will clear ambiguity on the nature of the queries from here on. 
There are also some TODOs that were addressed. 

It also partly implements what https://github.com/llvm/llvm-project/pull/87723 is attempting to accomplish. At least, on a point-to-point query between references, the distinction is made. To apply it to TBAA, would be another PR.

Note that, the changes were minimal in the TBAA code to retain the current results.

### CODE REVIEW

File: flang/lib/Optimizer/Transforms/AddAliasTags.cpp
Comment (vzakhari) [inline flang/lib/Optimizer/Transforms/AddAliasTags.cpp:166]: Is it possible to pass this check when the source is `Indirect` or `Allocate`?  I think that would be incorrect.  Can we remove this `else if` for now?  @tblah, do you agree?

**Review by vzakhari — COMMENTED** [review]
Thank you for the changes, Renaud!  It looks clearer to me now.  I have a question about the TBAA tags pass.

File: flang/lib/Optimizer/Transforms/AddAliasTags.cpp
Comment (Renaud-K) [reply flang/lib/Optimizer/Transforms/AddAliasTags.cpp:166]: Yes, this is not equivalent to what it was before.
This would be:
```
enableDirect && source.kind == SourceKind::Global && source.isBoxData()
```

Of course, if we remove SourceKind::Direct maybe the nomenclature for enableDirect could change as well. 
But the behavior should be equivalent.

Comment (tblah) [reply flang/lib/Optimizer/Transforms/AddAliasTags.cpp]: Thanks for spotting this Slava, yes you were right here and the update was needed.

My intention behind this `enableDirect` path was for certain spec benchmarks with a lot of boxed arrays as globals. I agree that some better naming to make this clearer.

**Review by tblah — COMMENTED** [review]
Thanks for this, it gives a big improvement to readability and it is good to catch more cases.

File: flang/lib/Optimizer/Analysis/AliasAnalysis.cpp
Comment (jeanPerier) [inline flang/lib/Optimizer/Analysis/AliasAnalysis.cpp:128]: I think this may not be correct with derived types containing POINTERs/ALLOCATABLES.

```
module m
type t
 type(t), pointer :: next
 integer :: i
end type
contains
subroutine foo(x, y)
  type(t) :: x, y
  integer :: i1, i2
  i1 = x%next%i
  x = y
  i2 = x%next%i
end subroutine
end module
```

An optimization trying to merge similar FIR loads would run into trouble if it relied on the alias analysis because it would incorrectly think it can merge the loads of the two `x%next` descriptor loads.

The simplified FIR looks like:

````
 !_QMmTt = !fir.type<_QMmTt{next:!fir.box<!fir.ptr<!fir.type<_QMmTt>>>,i:i32}>

func.func @_QMmPfoo(%x: !fir.ref<!_QMmTt> {fir.bindc_name = "x"}, %y: !fir.ref<!_QMmTt>) {
  %i1 = fir.alloca i32
  %i2 = fir.alloca i32
  %2 = hlfir.designate %x{"next"}   {fortran_attrs = #fir.var_attrs<pointer>} : (!fir.ref<!_QMmTt>) -> !fir.ref<!fir.box<!fir.ptr<!_QMmTt>>>
  %3 = fir.load %2 : !fir.ref<!fir.box<!fir.ptr<!_QMmTt>>>
  %4 = fir.box_addr %3 : (!fir.box<!fir.ptr<!_QMmTt>>) -> !fir.ptr<!_QMmTt>
  %5 = hlfir.designate %4{"i"}   : (!fir.ptr<!_QMmTt>) -> !fir.ref<i32>
  %6 = fir.load %5 : !fir.ref<i32>
  hlfir.assign %6 to %i1 : i32, !fir.ref<i32>
  hlfir.assign %y to %x : !fir.ref<!_QMmTt>, !fir.ref<!_QMmTt>
  %7 = fir.load %2 : !fir.ref<!fir.box<!fir.ptr<!_QMmTt>>>
  // ... assign to "%i" to i2
  return
}
````

My issue is that I think with the new approach, a pass relying on memory side effects + alias to merge/move fir.load would merge the two fir.load from %2 (%7 and %3) because the only  write effects in between are the two hlfir.assign. The assign to `i1` obviously do not affect %2, but the one to `%x` does, yet since I think `%x` and `%2` are considered to not alias (same source, but %x is "data", while `%2` is not), it would seem legal to merge them.

**Review by jeanPerier — COMMENTED** [review]
Thank you @Renaud-K , it is great to simplify the logic here! I think you need to be more restrictive with types containing pointer/allocatables components (I see there are already some restrictions using `isRecordWithPointerComponent`, but they do not cover some previous early return cases).

**Review by jdenny-ornl — COMMENTED** [review]
Thanks for working on this.

File: flang/include/flang/Optimizer/Analysis/AliasAnalysis.h
Comment (Renaud-K) [reply flang/include/flang/Optimizer/Analysis/AliasAnalysis.h:41]: Good catch. Thank you.

File: flang/include/flang/Optimizer/Analysis/AliasAnalysis.h
Comment (Renaud-K) [reply flang/include/flang/Optimizer/Analysis/AliasAnalysis.h:57]: Yes. I will comment this more, since you are all supportive of the idea. Thank you for that by the way.

File: flang/lib/Optimizer/Analysis/AliasAnalysis.cpp
Comment (Renaud-K) [reply flang/lib/Optimizer/Analysis/AliasAnalysis.cpp:128]: We have `isRecordWithPointerComponent` which is a pretty big hammer. It does not distinguish between data and non-data, so %x and %2 will alias. Also in this case, there are indirect loads. The source of %5 is %3 and we do not get passed this load because it is not a source.  I will add the test. There is room for improvement. For instance `isRecordWithPointerComponent` even says that x and y alias.

Comment (Renaud-K) [reply flang/lib/Optimizer/Analysis/AliasAnalysis.cpp:128]: This is a good catch and actually an easy fix. The catch all statement at the top is not needed:
I will be removing this:
```
  if ((lhsSrc.origin.u == rhsSrc.origin.u) &&
      lhsSrc.isData() != rhsSrc.isData()) {
```

Comment (jeanPerier) [reply flang/lib/Optimizer/Analysis/AliasAnalysis.cpp]: Fix looks great, thanks for adding a test!

**Review by jeanPerier — APPROVED** [review]
Thanks Renaud, please wait on further approval WRT the comments from other reviewers, but the code logic looks great to me.

**Review by tblah — APPROVED** [review]
LGTM once other reviewers are happy and CI passes. I don't see any performance regressions from a very non-exhaustive check.

Thank you for your work on this. I think it is a lot easier to understand than the older approach.

File: flang/test/Analysis/AliasAnalysis/alias-analysis-9.fir
Comment (Renaud-K) [reply flang/test/Analysis/AliasAnalysis/alias-analysis-9.fir:20]: The three.

File: flang/test/Analysis/AliasAnalysis/alias-analysis-9.fir
Comment (Renaud-K) [reply flang/test/Analysis/AliasAnalysis/alias-analysis-9.fir:28]: We save MustAlias when the starting points match exactly. If we reach the source through an offset calculation we set `approximateSource` and return MayAlias. https://github.com/llvm/llvm-project/commit/c3f060b026a1e7f4676038fffce5ee2bea1292ac


As far, `xnext1#0 <-> xnext2#0`, we could do better and return MustAlias but as @tblah points out in his commit, we do not compare operands used in offset calculation.

File: flang/lib/Optimizer/Analysis/AliasAnalysis.cpp
Comment (jdenny-ornl) [reply flang/lib/Optimizer/Analysis/AliasAnalysis.cpp:111]: This issue remains.

File: flang/test/Analysis/AliasAnalysis/alias-analysis-9.fir
Comment (jdenny-ornl) [inline flang/test/Analysis/AliasAnalysis/alias-analysis-9.fir:28]: For this comment to be relevant, `x#0` must be considered either a pointer component or pointer object.  But `x#0` is like `y#0`, which the previous block says is not a pointer object and does not alias `xnext1#0` or `xnext2#0`.  Please clarify.

**Review by jdenny-ornl — COMMENTED** [review]
Thanks for the updates.  Just a few more minor points.

File: flang/test/Analysis/AliasAnalysis/alias-analysis-9.fir
Comment (Renaud-K) [reply flang/test/Analysis/AliasAnalysis/alias-analysis-9.fir:28]: git-clang-format is moving the llvm::dbgs statement next to the first one. But I can
`llvm::dbgs() << "\nAliasAnalysis::alias\n";`

**Review by jdenny-ornl — APPROVED** [review]
LGTM once CI passes.  Thanks!

**Comment by Renaud-K** [comment]
I tried to commit to my fork a rebase of the branch. Sorry if it caused confusion.

**Pull request merged**
